File: VB/Scaffolding.ValidationErrors/Common/ViewModel/ReadOnlyCollectionViewModel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Mvvm.DataAnnotations
Imports Scaffolding.ValidationErrors.Common.Utils
Imports Scaffolding.ValidationErrors.Common.DataModel

Namespace Scaffolding.ValidationErrors.Common.ViewModel
	Public MustInherit Partial Class ReadOnlyCollectionViewModel(Of TEntity As Class, TUnitOfWork As IUnitOfWork)
		Inherits ReadOnlyCollectionViewModelBase(Of TEntity, TUnitOfWork)
		Public Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IReadOnlyRepository(Of TEntity)), Optional ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)) = Nothing)
			MyBase.New(unitOfWorkFactory, getRepositoryFunc, filterExpression)
		End Sub
	End Class
	<POCOViewModel> _
	Public MustInherit Class ReadOnlyCollectionViewModelBase(Of TEntity As Class, TUnitOfWork As IUnitOfWork)
		Private refreshOnFilterExpressionChanged As Boolean = False
		Private repository_Renamed As IReadOnlyRepository(Of TEntity)
		Protected ReadOnly unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork)
		Private ReadOnly getRepositoryFunc As Func(Of TUnitOfWork, IReadOnlyRepository(Of TEntity))
		Private entities_Renamed As IList(Of TEntity)

		Public Sub New(ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of TUnitOfWork), ByVal getRepositoryFunc As Func(Of TUnitOfWork, IReadOnlyRepository(Of TEntity)), ByVal filterExpression As Expression(Of Func(Of TEntity, Boolean)))
			Me.unitOfWorkFactory = unitOfWorkFactory
			Me.getRepositoryFunc = getRepositoryFunc
			FilterExpression = filterExpression
			Me.refreshOnFilterExpressionChanged = True
			If (Not Me.IsInDesignMode()) Then
				OnInitializeInRuntime()
			End If
		End Sub
		Protected Overridable Sub OnInitializeInRuntime()
		End Sub
		Public Overridable Sub Refresh()
			Me.repository_Renamed = GetRepository()
			Me.entities_Renamed = GetEntities()
			Me.RaisePropertyChanged(Function(x) Entities)
		End Sub

		Protected ReadOnly Property Repository() As IReadOnlyRepository(Of TEntity)
			Get
				If repository_Renamed Is Nothing Then
					repository_Renamed = GetRepository()
				End If
				Return repository_Renamed
			End Get
		End Property
		Public ReadOnly Property Entities() As IList(Of TEntity)
			Get
				If entities_Renamed Is Nothing Then
					entities_Renamed = GetEntities()
				End If
				Return entities_Renamed
			End Get
		End Property
		Private privateSelectedEntity As TEntity
		Public Overridable Property SelectedEntity() As TEntity
			Get
				Return privateSelectedEntity
			End Get
			Set(ByVal value As TEntity)
				privateSelectedEntity = value
			End Set
		End Property
		Protected Overridable Sub OnSelectedEntityChanged()
		End Sub
		Private privateFilterExpression As Expression(Of Func(Of TEntity, Boolean))
		Public Overridable Property FilterExpression() As Expression(Of Func(Of TEntity, Boolean))
			Get
				Return privateFilterExpression
			End Get
			Set(ByVal value As Expression(Of Func(Of TEntity, Boolean)))
				privateFilterExpression = value
			End Set
		End Property
		Protected Overridable Sub OnFilterExpressionChanged()
			If refreshOnFilterExpressionChanged Then
				Refresh()
			End If
		End Sub
		Private Function GetRepository() As IReadOnlyRepository(Of TEntity)
			Return getRepositoryFunc(unitOfWorkFactory.CreateUnitOfWork())
		End Function
		Protected Overridable Function GetEntities() As IList(Of TEntity)
			Dim queryable = Repository.GetEntities()
			If FilterExpression IsNot Nothing Then
				queryable = queryable.Where(FilterExpression)
			End If
			queryable.Load()
			Return Repository.Local
		End Function
	End Class
End Namespace

File: CS/Scaffolding.ValidationErrors/Common/ViewModel/ReadOnlyCollectionViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.DataAnnotations;
using Scaffolding.ValidationErrors.Common.Utils;
using Scaffolding.ValidationErrors.Common.DataModel;

namespace Scaffolding.ValidationErrors.Common.ViewModel {
    public abstract partial class ReadOnlyCollectionViewModel<TEntity, TUnitOfWork> : ReadOnlyCollectionViewModelBase<TEntity, TUnitOfWork>
        where TEntity : class
        where TUnitOfWork : IUnitOfWork {
        public ReadOnlyCollectionViewModel(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc, Expression<Func<TEntity, bool>> filterExpression = null)
            : base(unitOfWorkFactory, getRepositoryFunc, filterExpression) {
        }
    }
    [POCOViewModel]
    public abstract class ReadOnlyCollectionViewModelBase<TEntity, TUnitOfWork>
        where TEntity : class
        where TUnitOfWork : IUnitOfWork {
        bool refreshOnFilterExpressionChanged = false;
        IReadOnlyRepository<TEntity> repository;
        protected readonly IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory;
        readonly Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc;
        IList<TEntity> entities;

        public ReadOnlyCollectionViewModelBase(IUnitOfWorkFactory<TUnitOfWork> unitOfWorkFactory, Func<TUnitOfWork, IReadOnlyRepository<TEntity>> getRepositoryFunc, Expression<Func<TEntity, bool>> filterExpression) {
            this.unitOfWorkFactory = unitOfWorkFactory;
            this.getRepositoryFunc = getRepositoryFunc;
            FilterExpression = filterExpression;
            this.refreshOnFilterExpressionChanged = true;
            if(!this.IsInDesignMode())
                OnInitializeInRuntime();
        }
        protected virtual void OnInitializeInRuntime() { }
        public virtual void Refresh() {
            this.repository = GetRepository();
            this.entities = GetEntities();
            this.RaisePropertyChanged(x => Entities);
        }

        protected IReadOnlyRepository<TEntity> Repository {
            get {
                if(repository == null)
                    repository = GetRepository();
                return repository;
            }
        }
        public IList<TEntity> Entities {
            get {
                if(entities == null)
                    entities = GetEntities();
                return entities;
            }
        }
        public virtual TEntity SelectedEntity { get; set; }
        protected virtual void OnSelectedEntityChanged() { }
        public virtual Expression<Func<TEntity, bool>> FilterExpression { get; set; }
        protected virtual void OnFilterExpressionChanged() {
            if(refreshOnFilterExpressionChanged)
                Refresh();
        }
        IReadOnlyRepository<TEntity> GetRepository() {
            return getRepositoryFunc(unitOfWorkFactory.CreateUnitOfWork());
        }
        protected virtual IList<TEntity> GetEntities() {
            var queryable = Repository.GetEntities();
            if(FilterExpression != null)
                queryable = queryable.Where(FilterExpression);
            queryable.Load();
            return Repository.Local;
        }
    }
}